sql/dd/impl/bootstrap/bootstrapper.cc

```diff
@@ -1341,7 +1354,25 @@ bool initialize_dd_properties(THD *thd) {
       return true;
     }
     bootstrap::DD_bootstrap_ctx::instance().set_actual_dd_engine(
-        static_cast<legacy_db_type>(actual_dd_engine));
+        actual_dd_engine);
+
+    /*
+      Read DD_SYSTEM_ENGINE_UPGRADED from dd_properties, if not exist,
+      use INNODB instead;
+    */
+    legacy_db_type upgraded_dd_system_engine = DB_TYPE_UNKNOWN;
+    bool exists_upgraded_dd_system_engine = false;
+    if (dd::tables::DD_properties::instance().get(
+            thd, "DD_SYSTEM_ENGINE_UPGRADED",
+            reinterpret_cast<uint *>(&upgraded_dd_system_engine),
```

Comment:
Likewise

Reply:
same as before, static_cast doesn't allow this.